File: 79-word-search/79-word-search.py
class Solution:
    
    def solve(self, board, word, x, i, j):
        
        if (x >= len(word)):
            return True
        
        if not (0 <= i < len(board) and 0 <= j < len(board[0])):
            return False
        
        if (board[i][j] == word[x]):
            board[i][j] = "-1"
            solved = self.solve(board, word, x + 1, i, j + 1) or \
                    self.solve(board, word, x + 1, i, j - 1) or \
                    self.solve(board, word, x + 1, i + 1, j) or \
                    self.solve(board, word, x + 1, i - 1, j)
            board[i][j] = word[x]
            return solved
        return False
        
    
    def exist(self, board: List[List[str]], word: str) -> bool:
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.solve(board, word, 0, i, j):
                    return True
        return False

File: 79-word-search/79-word-search.cpp
class Solution {
public:
    bool solve(vector<vector<char>>& board, string word, int x, int i, int j)
    {
        if (x == word.size())
            return true;
        
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size())
            return false;
        
        bool solved = false;
        
        if (word[x] == board[i][j])
        {
            board[i][j] = '-';
            solved = solve(board, word, x + 1, i + 1, j) || 
                solve(board, word, x + 1, i - 1, j) ||
                solve(board, word, x + 1, i, j - 1) ||
                solve(board, word, x + 1, i, j + 1);
            board[i][j] = word[x];
        }
        return solved;
    }
    
    bool exist(vector<vector<char>>& board, string word) 
    {
        for (int i=0; i<board.size(); i++)
        {
            for (int j=0; j<board[0].size(); j++)
            {
                if (solve(board, word, 0, i, j))
                    return true;
            }
        }
        return false;
    }
};
